### PR TITLE
Add missing common mathematical function (absolute value)

### PR DESCRIPTION
**Pull request by dalg24**
Partially addressing #4033

### CODE REVIEW

File: core/unit_test/TestMathematicalFunctions.hpp [hunk 1]
```diff
@@ -867,3 +867,44 @@ TEST(TEST_CATEGORY,
...
+      KOKKOS_IMPL_DO_NOT_USE_PRINTF("failed abs(int)\n");
+    }
+    if (abs(2.) != 2. || abs(-2.) != 2.) {
+      ++e;
+      KOKKOS_IMPL_DO_NOT_USE_PRINTF("failed abs(double)\n");
+    }
+    // special values
+    using Kokkos::Experimental::isinf;
+    using Kokkos::Experimental::isnan;
+    if (abs(-0.) != 0. || !isinf(abs(-INFINITY)) || !isnan(-NAN)) {
```
Comment (dalg24) [inline core/unit_test/TestMathematicalFunctions.hpp:892]:
```suggestion
    if (abs(-0.) != 0. || !isinf(abs(-INFINITY)) || !isnan(abs(-NAN))) {
```

File: core/src/Kokkos_MathematicalFunctions.hpp [hunk 2]
```diff
@@ -227,6 +227,25 @@ KOKKOS_IMPL_MATH_UNARY_PREDICATE(isnan)
...
+#ifdef KOKKOS_ENABLE_SYCL
+KOKKOS_INLINE_FUNCTION int abs(int n) { return sycl::abs(n); }
+KOKKOS_INLINE_FUNCTION long abs(long n) { return sycl::abs(n); }
+KOKKOS_INLINE_FUNCTION long long abs(long long n) { return sycl::abs(n); }
+KOKKOS_INLINE_FUNCTION float abs(float x) { return sycl::fabs(x); }
+KOKKOS_INLINE_FUNCTION double abs(double x) { return sycl::fabs(x); }
+#else
+KOKKOS_INLINE_FUNCTION int abs(int n) { using std::abs;  return abs(n); }
+KOKKOS_INLINE_FUNCTION long abs(long n) { using std::labs;  return labs(n); }
+KOKKOS_INLINE_FUNCTION long long abs(long long n) { using std::llabs;  return llabs(n); }
```
Comment (nliber) [inline core/src/Kokkos_MathematicalFunctions.hpp:241]: While I don't think it matters in this case, there are overloads for `std::abs` for `long`, `long long`, `float`, `double` and `long double`.

**Review by nliber — APPROVED** [review]

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/Kokkos_MathematicalFunctions.hpp:241]: I am trying to find something that works with the CUDA math functions and these overloads are missing unfortunately

Comment (dalg24) [reply core/src/Kokkos_MathematicalFunctions.hpp:241]: CUDA provides the following (unqualified) functions
```C++
__device__ ​ int abs ( int  a );
__device__ ​ long int labs ( long int  a );
__device__ ​ long long int 	llabs ( long long int a );
__device__ ​ float fabsf ( float  x );
__device__ ​ double fabs ( double  x );

File: core/unit_test/TestMathematicalFunctions.hpp [hunk 3]
```diff
@@ -867,3 +867,62 @@ TEST(TEST_CATEGORY,
...
+#ifdef MATHEMATICAL_FUNCTIONS_HAVE_LONG_DOUBLE_OVERLOADS
+    if (abs(6.l) != 6.l || abs(-6.l) != 6.l) {
+      ++e;
+      KOKKOS_IMPL_DO_NOT_USE_PRINTF("failed abs(long double)\n");
+    }
+#endif
+    // special values
+    using Kokkos::Experimental::isinf;
+    using Kokkos::Experimental::isnan;
+    if (abs(-0.) != 0. || !isinf(abs(-INFINITY)) || !isnan(abs(-NAN))) {
```
Comment (dalg24) [inline core/unit_test/TestMathematicalFunctions.hpp:910]: Failed on some of the builds
https://github.com/kokkos/kokkos/pull/4036/checks?check_run_id=2622966686#step:7:1253

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 2 above)
Comment (nliber) [reply core/src/Kokkos_MathematicalFunctions.hpp]: Got it.  Thanks for the explanation.

File: core/unit_test/TestMathematicalFunctions.hpp (hunk 3 above)
Comment (dalg24) [reply core/unit_test/TestMathematicalFunctions.hpp:910]: With the help from @jrmadsen and @masterleinad 
https://godbolt.org/z/EdYPsaxM8 ugggggg

Comment (dalg24) [reply core/unit_test/TestMathematicalFunctions.hpp]: I decided to disable when icpx is used

**Comment by masterleinad** [comment]
I have a slight preference towards something like
```diff
diff --git a/.github/workflows/continuous-integration-workflow.yml b/.github/workflows/continuous-integration-workflow.yml
index 892ca127a..c2ce93235 100644
--- a/.github/workflows/continuous-integration-workflow.yml
+++ b/.github/workflows/continuous-integration-workflow.yml
@@ -45,6 +45,13 @@ jobs:
       - name: maybe_disable_death_tests
         if: ${{ matrix.distro == 'fedora:rawhide' }}
         run: echo "GTEST_FILTER=-*DeathTest*" >> $GITHUB_ENV
+      - name: enable_precise_fp_mode
+        run: |
+          if ${{ matrix.cxx == 'icpx' }}; then
+            echo "EXTRA_CXX_FLAGS=-fp-model=precise" >> $GITHUB_ENV
+          else
+            echo "EXTRA_CXX_FLAGS=" >> $GITHUB_ENV
+          fi
       - name: build-and-test
         run: |
           ccache -z
@@ -56,6 +63,7 @@ jobs:
             -DKokkos_ENABLE_EXAMPLES=ON \
             -DKokkos_ENABLE_DEPRECATED_CODE_3=ON \
             -DCMAKE_CXX_COMPILER=${{ matrix.cxx }} \
+            -DCMAKE_CXX_FLAGS=${EXTRA_CXX_FLAGS} \
             -DCMAKE_BUILD_TYPE=${{ matrix.cmake_build_type }} \
             -DBUILD_NAME=${{ matrix.distro }}-${{ matrix.cxx }} \
             -DBUILD_JOBS=2 -DBINARY_DIR=builddir -DSITE=GitHub-Linux \
```
see https://github.com/masterleinad/kokkos/runs/2640376010?check_suite_focus=true saying "We only test in floating point safe mode. You are on your own if you use unsafe floating point operations (that happen to be enabled by default in certain `Intel` compilers)."
I can live with either, though.

**Review by masterleinad — APPROVED** [review]

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

File: core/src/Kokkos_MathematicalFunctions.hpp (hunk 2 above)
Comment (dalg24) [reply core/src/Kokkos_MathematicalFunctions.hpp]: I was wrong.  I based my answer on the documented function in the [CUDA math API](https://docs.nvidia.com/cuda/cuda-math-api/modules.html#modules) but I checked  the hardware counters with `nvprof` and `::abs((int)i)` does not give any floating-point instruction.

**Review by crtrott — APPROVED** [review]

**Pull request merged**